### DISCUSSION TITLE
Transfer/connect the state to an existing state management (react redux toolkit)

### DISCUSSION THREAD

**Discussion by AsifDaum** [xyflow/xyflow#1009] [Q&A]
Hi, apologies in advance as I am fairly new so my skills arent that great. 
I have been trying to connect the store to my existing toolkit the `useStoreState` acts as a `useSelector` it isnt ideal for my use case. I am wondering if there is a way to either connect the store to my existing redux toolkit store or whether I would have to dispatch to the store.

My current issue is that I am trying to place the active `elements` into my store and have each object linked another item, however since `useStoreState` returns the entire object which in turn becomes non-serializable when I `dispatch` it into the store. This means that whenever an element is deleted/added it refreshes that stored state in the redux toolkit store, which means all the linked objects lose their connection and become deleted as well.

Is there any possible work around here that I could look at? Perhaps being able to store the results directly into the existing redux toolkit store somehow? I know easy peasy is used, and I tried researching on how to connect it to the store which I failed in achieving too.

`export default connect((state) => ({ elements: state}))(EditorReactFlow);`

Am I also supposed to see the internal store in the redux devtools?

Another thing I noticed is that when I used the `useStoreState` that it would mount my component thrice on any changes to the elements/store? It didnt happen when I simply dispatched the elements using a  

```
useEffect(() =>{
    dispatch(addStep(elements))
  },[dispatch, elements])
```
That would mount it once each time the elements changed. I will investigate this further, but its something interesting I found.

**Comment by harshagarwalnyu** [comment]
Use **controlled mode**: store `nodes` and `edges` in Redux, use React Flow's change helpers to compute next state.

```tsx
// store/flowSlice.ts
import { createSlice } from '@reduxjs/toolkit';
import { applyNodeChanges, applyEdgeChanges } from 'reactflow';

const flowSlice = createSlice({
  name: 'flow',
  initialState: { nodes: [], edges: [] },
  reducers: {
    nodesChanged: (state, action) => {
      state.nodes = applyNodeChanges(action.payload, state.nodes);
    },
    edgesChanged: (state, action) => {
      state.edges = applyEdgeChanges(action.payload, state.edges);
    },
    setNodes: (state, action) => { state.nodes = action.payload; },
    setEdges: (state, action) => { state.edges = action.payload; },
  },
});

export const { nodesChanged, edgesChanged, setNodes, setEdges } = flowSlice.actions;
```

```tsx
// FlowChart.tsx
import ReactFlow, { addEdge } from 'reactflow';
import { useSelector, useDispatch } from 'react-redux';

export default function FlowChart() {
  const nodes = useSelector((s) => s.flow.nodes);
  const edges = useSelector((s) => s.flow.edges);
  const dispatch = useDispatch();

  return (
    <ReactFlow
      nodes={nodes}
      edges={edges}
      onNodesChange={(changes) => dispatch(nodesChanged(changes))}
      onEdgesChange={(changes) => dispatch(edgesChanged(changes))}
      onConnect={(params) => dispatch(setEdges(addEdge(params, edges)))}
    />
  );
}
```

**On the non-serializable warning:** React Flow node/edge objects (`{ id, position, data, type, selected, dragging }`) are fully serializable. The warning is likely from a callback stored inside `node.data` — move callbacks out of node data and look them up by ID instead:

```ts
// Don't: node.data = { onClick: () => {...} }
// Do: node.data = { id: 'node-1' }
// Then: const handler = handlers[node.data.id]
```

This pattern gives you full Redux DevTools time-travel, undo/redo, and cross-component access to all flow state.
